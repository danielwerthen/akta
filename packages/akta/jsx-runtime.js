export { Fragment, jsx, jsxs } from './dist';
