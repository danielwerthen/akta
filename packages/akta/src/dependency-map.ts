import { BehaviorSubject, Observable, of } from 'rxjs';
import { switchMap } from 'rxjs/operators';

type EagerDependency<T> = {
  key: symbol;
  value: T;
};
type LazyDependency<T> = {
  key: symbol;
  init: () => T;
};

export type Dependency<T> = EagerDependency<T> | LazyDependency<T>;

export function isLazyDependency<T>(
  dep: Dependency<T>
): dep is LazyDependency<T> {
  return typeof (dep as LazyDependency<T>).init === 'function';
}

const LocalStop = Symbol('Dependency stopper');

export function getInitialValue<T>(dep: Dependency<T>): T {
  if (isLazyDependency(dep)) {
    return dep.init();
  }
  return dep.value;
}

export function createDependency<T>(initialValue: T): Dependency<T> {
  return {
    key: Symbol('Dependency'),
    value: initialValue,
  };
}

export function createLazyDependency<T>(init: () => T): Dependency<T> {
  return {
    key: Symbol('Dependency'),
    init,
  };
}

export class DependencyMap {
  private store: Map<Symbol, BehaviorSubject<unknown>> = new Map();
  parent?: DependencyMap;
  constructor(parent?: DependencyMap) {
    this.parent = parent;
  }

  private getLocal<T>(dep: Dependency<T>): BehaviorSubject<T> {
    const { key } = dep;
    const subject = this.store.get(key) ?? new BehaviorSubject(LocalStop);
    if (!this.store.has(key)) {
      this.store.set(key, subject);
    }
    return subject as BehaviorSubject<T>;
  }
  observe<T>(dep: Dependency<T>): Observable<T> {
    const subject = this.getLocal(dep);
    return subject.pipe(
      switchMap(val => {
        if ((val as unknown) === LocalStop) {
          return this.parent
            ? this.parent.observe(dep)
            : of(getInitialValue(dep));
        }
        return of(val);
      })
    );
  }
  peek<T>(dep: Dependency<T>): T {
    const value = this.getLocal(dep).value as unknown;
    if (value !== LocalStop) {
      return value as T;
    }
    if (this.parent) {
      return this.parent.peek(dep);
    }
    return getInitialValue(dep);
  }
  branch(): DependencyMap {
    return new DependencyMap(this);
  }
  provide<T>(dep: Dependency<T>, value: T | ((old?: T) => T)) {
    const subject = this.getLocal(dep);
    if (typeof value === 'function') {
      const fn = value as <T>(old?: T) => T;
      subject.next(fn(this.peek(dep) ?? undefined));
    } else {
      subject.next(value);
    }
  }
  unprovide(dep: Dependency<unknown>): unknown {
    const subject = this.getLocal(dep);
    const prev = subject.value;
    if (prev !== LocalStop) {
      subject.next(LocalStop);
      return prev;
    }
    return;
  }
}
