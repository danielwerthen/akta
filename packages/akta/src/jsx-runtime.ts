// Based on https://raw.githubusercontent.com/DefinitelyTyped/DefinitelyTyped/314d0c3cf25e1762525a3790b4609eb19fddadf9/types/react/index.d.ts

import { AktaElement, AktaElementType } from './types';

export function jsx<
  P extends {},
  T extends string | ((props: P) => AktaElement<P, any>)
>(type: T, props: P, maybeKey?: string): AktaElement<P, T> {
  return {
    $$type: AktaElementType,
    type,
    props,
    key: maybeKey,
  };
}

export const jsxs = jsx;
