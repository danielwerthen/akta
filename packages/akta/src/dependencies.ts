import { BehaviorSubject, firstValueFrom, Subject, switchMap } from 'rxjs';
import { createDependency, DependencyMap } from './dependency-map';
import { AllElements, isEvent } from './element-ops';
import { lazy } from './lazy-function';
import MetaObject, { MethodMissing } from './meta-object';
import { createSyncContext } from './sync-context';
import { AktaNode, ElementProperties } from './types';

export const dependecyContext = createSyncContext<DependencyMap>();

export const continuationDependency = createDependency(lazy());
export type TeardownFunction = () => void;
export const teardownDependency = createDependency<null | TeardownFunction[]>(
  null
);

export const elementsDependency = createDependency(new AllElements());

export function useNext<T extends AktaNode, RETURN extends AktaNode, NEXT>(
  _input?: () => Generator<T, RETURN, NEXT> | AsyncGenerator<T, RETURN, NEXT>
): (input: NEXT) => Promise<void> {
  const ctx = dependecyContext.getContext();
  return (input: NEXT) => {
    return firstValueFrom(
      ctx(continuationDependency).pipe(switchMap(next => next(input)))
    ).then(() => void 0);
  };
}

export function useTeardown(fn: TeardownFunction) {
  const ctx = dependecyContext.getContext();
  ctx.provide(teardownDependency, fns => {
    if (!fns) {
      return [fn];
    }
    fns.push(fn);
    return fns;
  });
}

export class PropertiesBase extends MetaObject<
  | ((factory?: () => Subject<unknown>) => Subject<unknown>)
  | ((start?: unknown) => Subject<unknown>)
> {
  [MethodMissing](key: string) {
    if (isEvent(key)) {
      return (factory?: () => Subject<unknown>) =>
        factory ? factory() : new Subject();
    } else {
      return (initialValue?: unknown) => {
        if (initialValue === undefined) {
          return new Subject();
        } else {
          return new BehaviorSubject(initialValue);
        }
      };
    }
  }
}

const propertiesBase = new PropertiesBase();
export class ElementPropertiesBase extends MetaObject<PropertiesBase> {
  [MethodMissing](_key: string) {
    return propertiesBase;
  }
}

export const Props: ElementProperties = (new ElementPropertiesBase() as unknown) as ElementProperties;
