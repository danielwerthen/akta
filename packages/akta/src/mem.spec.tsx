import { Subject } from 'rxjs';
import { dependecyContext } from './dependencies';
import { DependencyMap } from './dependency-map';
import { mount } from './';
import { jsx } from './jsx-runtime';

describe('ObserveAsync', () => {
  it('Array nodes are rendered correctly', async () => {
    const root = document.createElement('div');
    function Comp() {
      return jsx('p', { children: ['test', 'less'] });
    }
    const value = new Subject();

    dependecyContext.setContext(
      () => mount(jsx('div', { children: value }), root),
      new DependencyMap()
    );
    value.next([jsx(Comp, {})]);
    expect(root).toMatchSnapshot();
  });
});
