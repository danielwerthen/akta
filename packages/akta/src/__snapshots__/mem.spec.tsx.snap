// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ObserveAsync Array nodes are rendered correctly 1`] = `
<div>
  <div>
    <p>
      test
      less
    </p>
  </div>
</div>
`;
